services/src/machine_learning/mod.rs

```diff
@@ -0,0 +1,106 @@
+mod data_preparation;
+mod ml_tasks;
+pub mod xg_error;
+mod xgboost_training;
+
+pub(crate) use ml_tasks::{schedule_ml_model_training_task, MLTrainRequest};
+
+#[cfg(test)] //TODO: remove test config, once its used outside of tests
+pub(crate) use ml_tasks::MachineLearningModelFromWorkflowResult;
+
+use serde::Deserialize;
+use serde::Serialize;
+use typetag::serde;
+pub(crate) use xgboost_training::xgb_train_model;
+pub(crate) use xgboost_training::XgboostTrainingParams;
+
+/// Represents a single feature of a machine learning training data set
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct MachineLearningFeature {
+    pub feature_name: Option<String>, // e.g. temperature, elevation etc.
+    pub feature_data: Vec<f32>,
+}
+
+impl MachineLearningFeature {
+    pub fn new(feature_name: Option<String>, feature_data: Vec<f32>) -> MachineLearningFeature {
+        MachineLearningFeature {
+            feature_name,
+            feature_data,
+        }
+    }
+}
+
+/// This enum represents the different aggregators that can be used to initialize the
+/// different algorithms for collecting the data used in ml training.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub enum MachineLearningAggregator {
+    Simple,
+    ReservoirSampling,
+}
+
+/// The purpose of this trait is to facilitate a generic aggregation process of raster data.
+pub trait Aggregatable {
+    type Data;
+
+    /// This method should realize the aggregation algorithm of the implementing struct.
+    fn aggregate(&mut self, incoming_data: &mut dyn Iterator<Item = Option<f32>>);
+
+    /// Once the aggregation process is finished, return the data for further usage.
+    fn finish(&self) -> &Self::Data;
+
+    /// Used to generate a new instance of an aggregatable within a generic context
+    fn new(data: Vec<f32>, memory_limit: Option<usize>) -> Self;
+}
+
+/// A simple aggregator that just collects all incoming data in a vector.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct SimpleAggregator {
+    pub data: Vec<f32>,
+}
+
+impl Aggregatable for SimpleAggregator {
+    type Data = Vec<f32>;
+
+    fn aggregate(&mut self, incoming_data: &mut dyn Iterator<Item = Option<f32>>) {
+        let v = incoming_data.into_iter().flatten().collect::<Vec<f32>>();
+
+        self.data.extend(v);
+    }
+
+    fn finish(&self) -> &Vec<f32> {
```

Comment:
could either return a `&[f32]` or consume self (change `&self` to `self`) and move the data out of the aggregator

Reply:
c4ee563f

commit msg:
 refactor: change Aggregatable::finish to return owned data and change the associated type